server-common/src/main/java/org/apache/kafka/server/group/share/PersisterStateManager.java

```diff
@@ -292,6 +305,79 @@ protected void findCoordinatorErrorResponse(Errors error, Exception exception) {
+      return new ReadShareGroupStateRequest.Builder(new ReadShareGroupStateRequestData()
+          .setGroupId(groupId)
+          .setTopics(Collections.singletonList(
+              new ReadShareGroupStateRequestData.ReadStateData()
+                  .setTopicId(topicId)
+                  .setPartitions(Collections.singletonList(
+                      new ReadShareGroupStateRequestData.PartitionData()
+                          .setPartition(partition)
+                          .setLeaderEpoch(leaderEpoch))))));
+    }
+
+    @Override
+    protected boolean isRequestResponse(ClientResponse response) {
+      return response.requestHeader().apiKey() == ApiKeys.READ_SHARE_GROUP_STATE;
+    }
+
+    @Override
+    protected void handleRequestResponse(ClientResponse response) {
+      log.info("Read state response received. - {}", response);
+      ReadShareGroupStateResponseData readShareGroupStateResponseData = ((ReadShareGroupStateResponse) response.responseBody()).data();
+      String errorMessage = "Failed to read state for partition " + partition + " in topic " + topicId + " for group " + groupId;
+      if (readShareGroupStateResponseData.results().size() != 1) {
+        log.error("ReadState response for {} is invalid", coordinatorKey);
+        this.future.complete(new ReadShareGroupStateResponse(
+            ReadShareGroupStateResponse.getErrorResponseData(topicId, partition, Errors.forException(new IllegalStateException(errorMessage)), errorMessage)));
+      }
+      ReadShareGroupStateResponseData.ReadStateResult topicData = readShareGroupStateResponseData.results().get(0);
+      if (!topicData.topicId().equals(topicId)) {
+        log.error("ReadState response for {} is invalid", coordinatorKey);
+        this.future.complete(new ReadShareGroupStateResponse(
+            ReadShareGroupStateResponse.getErrorResponseData(topicId, partition, Errors.forException(new IllegalStateException(errorMessage)), errorMessage)));
+      }
+      if (topicData.partitions().size() != 1) {
+        log.error("ReadState response for {} is invalid", coordinatorKey);
+        this.future.complete(new ReadShareGroupStateResponse(
+            ReadShareGroupStateResponse.getErrorResponseData(topicId, partition, Errors.forException(new IllegalStateException(errorMessage)), errorMessage)));
+      }
+      ReadShareGroupStateResponseData.PartitionResult partitionResponse = topicData.partitions().get(0);
+      if (partitionResponse.partition() != partition) {
+        log.error("ReadState response for {} is invalid", coordinatorKey);
+        this.future.complete(new ReadShareGroupStateResponse(
+            ReadShareGroupStateResponse.getErrorResponseData(topicId, partition, Errors.forException(new IllegalStateException(errorMessage)), errorMessage)));
+      }
+      future.complete((ReadShareGroupStateResponse) response.responseBody());
+    }
+
+    @Override
+    protected void findCoordinatorErrorResponse(Errors error, Exception exception) {
```

Comment:
could we simplify this method to:
```
@Override
    protected void findCoordinatorErrorResponse(Errors error, Exception exception) {
this.future.complete(new ReadShareGroupStateResponse(
            ReadShareGroupStateResponse.getErrorResponseData(topicId, partition, error, "Error in find coordinator. " +
                (exception == null ? error.message() : exception.getMessage()))));
}
```

Please also update the corresponding method for write handler

Reply:
Thanks for reviewing the PR, made the change in the latest commit